File: src/main/scala/de/theunknownxy/jgui/container/MultiContainer.scala
package de.theunknownxy.jgui.container

import de.theunknownxy.jgui.Widget
import de.theunknownxy.jgui.base.Point
import de.theunknownxy.jgui.event.MouseButton.MouseButton
import de.theunknownxy.jgui.layout.Constraint

import scala.collection.mutable

abstract class MultiContainer extends Container {
  var children: mutable.HashMap[Widget, Constraint] = mutable.HashMap()

  def addChild(widget: Widget, constraint: Constraint): Unit = {
    children += (widget -> constraint)
  }

  def removeChild(widget: Widget): Unit = {
    children -= widget
  }

  override def mouseClick(pos: Point, button: MouseButton): Option[Widget] = {
    children find(_._1.area contains pos) flatMap(_._1.mouseClick(pos, button))
  }

  override def draw(): Unit = {
    children map(_._1 draw())
  }
}

File: src/main/kotlin/de/theunknownxy/jgui/container/MultiContainer.kt
package de.theunknownxy.jgui.container

import de.theunknownxy.jgui.base.Point
import de.theunknownxy.jgui.event.MouseButton
import de.theunknownxy.jgui.base.Widget
import java.util.ArrayList

public abstract class MultiContainer : Container() {
    // TODO: Override width setter so that width == children.max(fixed? -> width)
    public var children: MutableList<Widget> = ArrayList()

    override fun draw() = children.forEach { w -> w.draw() }
    override fun onMouseClick(pos: Point, button: MouseButton): Widget? = children.firstOrNull { w -> w.rect.contains(pos) }?.onMouseClick(pos, button)
    override fun onUpdate() = children.forEach { w -> w.onUpdate() }
}